src/cursor.rs

```diff
@@ -0,0 +1,167 @@
+use std::{
+    collections::VecDeque,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use bson::{doc, Document};
+use derivative::Derivative;
+use futures::{future::BoxFuture, stream::Stream};
+
+use crate::{
+    error::Result,
+    operation::GetMore,
+    options::StreamAddress,
+    results::GetMoreResult,
+    Client,
+    Namespace,
+    RUNTIME,
+};
+
+#[derive(Debug)]
+pub(crate) struct CursorSpecification {
+    pub(crate) ns: Namespace,
+    pub(crate) address: StreamAddress,
+    pub(crate) id: i64,
+    pub(crate) batch_size: Option<u32>,
+    pub(crate) max_time: Option<Duration>,
+    pub(crate) buffer: VecDeque<Document>,
+}
+
+#[derive(Debug)]
+pub struct Cursor {
+    client: Client,
+    get_more: GetMore,
+    exhausted: bool,
+    state: State,
+}
+
+type GetMoreFuture = BoxFuture<'static, Result<GetMoreResult>>;
+
+/// Describes the current state of the Cursor. If the state is Executing, then a getMore operation
+/// is in progress. If the state is Buffer, then there are documents available from the current
+/// batch.
+#[derive(Derivative)]
+#[derivative(Debug)]
+enum State {
+    Executing(#[derivative(Debug = "ignore")] GetMoreFuture),
+    Buffer(VecDeque<Document>),
+    Exhausted,
+}
+
+impl Cursor {
+    pub(crate) fn new(client: Client, spec: CursorSpecification) -> Self {
+        let get_more = GetMore::new(
+            spec.ns,
+            spec.id,
+            spec.address,
+            spec.batch_size,
+            spec.max_time,
+        );
+
+        Self {
+            client,
+            get_more,
+            exhausted: spec.id == 0,
+            state: State::Buffer(spec.buffer),
+        }
+    }
+}
+
+impl Drop for Cursor {
+    fn drop(&mut self) {
+        if self.exhausted {
+            return;
+        }
+
+        let namespace = self.get_more.namespace().clone();
+        let client = self.client.clone();
+        let cursor_id = self.get_more.cursor_id();
+
+        RUNTIME.execute(async move {
+            let _: Result<_> = client
+                .database(&namespace.db)
+                .run_command(
+                    doc! {
+                        "killCursors": &namespace.coll,
+                        "cursors": [cursor_id]
+                    },
+                    None,
+                )
+                .await;
+        })
+    }
+}
+
+impl Stream for Cursor {
+    type Item = Result<Document>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        loop {
```

Comment:
The body of this method is the most significant change from the previous implementation. The issue was caused by us not polling the new future we created for the new getMore operation after exhausting a buffer, which meant that the context would not be passed through to the underlying calls to poll on the TcpSocket, so the task wouldn't be woken up when the response was ready. More generally, it's a mistake to return `Poll::Pending` without polling the underlying future. Because the logic for handling the result of polling the future needs to be the same as in the case that the state is `Executing`, the simplest way to implement it is to wrap the state checking in a loop. In the case that the buffer is empty but the cursor is not exhausted, we start a new getMore operation and then continue looping, which will cause the future to be polled immediately.